### ISSUE TITLE
Relate does not handle exceptions

### ISSUE THREAD

**Issue by mwtoews** [Toblerity/Shapely#294]
Here's an example with an invalid Polygon:
```python
from shapely.wkt import loads as load_wkt
a = load_wkt(
    'POLYGON ((40 100, 80 100, 80 60, 40 60, 40 100), (60 60, 80 60, 80 40, 60 40, 60 60))')
a.relate(a)  # oh no!
```
on Windows with Shapely 1.5.9, GEOS 3.4.2-CAPI-1.8.2 r3921, it shows

> Traceback (most recent call last):
> WindowsError: exception: access violation reading 0x000000000000000F

and on Debian with Shapely 1.5.8, GEOS 3.3.3-CAPI-1.7.4, it is dead with Segmentation fault.

However, it should probably raise a TopologicalError, like `a.touches(a)` does:
> shapely.geos.TopologicalError: The operation 'GEOSTouches_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x7fb1e5e4f090>

I should also note that I've checked all the other binary operators, and most of them raise TopologicalError, except for `distance`, which is 0.0, and that's ok. So only `relate` is at fault.

**Comment by snorfalorpagus** [comment]
It looks like GEOSRelate returns a NULL pointer on exception, which we don't check for.

https://github.com/libgeos/libgeos/blob/svn-trunk/capi/geos_c.h.in#L852

I think we need to add a check for this in `errcheck_just_free`.

**Comment by mwtoews** [comment]
The issue is effectively `string_at(None)`, which I've [reported upstream](http://bugs.python.org/issue24714).

And yes, part of the fix is in `errcheck_just_free`, to check `if result.value is None`. But then somehow raise a TopologicalError for relate.

**Comment by sgillies** [comment]
Closed via #303 and #304.
